src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -259,6 +259,15 @@ Kafka::Kafka(
 
     calculateDataFormat(storage);
 
+    /// For raw format, one_message_per_row default to true. Because if multiple rows are combined together,
+    /// it could be hard to separate them again.
+    if (data_format == "RawBLOB" && !settings->isChanged("one_message_per_row"))
+        settings->set("one_message_per_row", true);
+
+    support_count_optimization = data_format == "RawBLOB" ||
```

Comment:
In fact, we can't do the optimization for most formats, because most formats actually decode multiple rows from one single message. You can take a look at this page for the details: https://clickhouse.com/docs/en/interfaces/formats.

`one_message_per_row` is for sink, it only control how the external stream writes to kafka, but there are many other ways to write to a Kafka topic. So we can't decide if the optimization is doable by checking `one_message_per_row`. Because we don't have control on what messages are in a Kafka topic. For example, one can use benthos (now redpanda connect) to generate Kafka messages that contains multiple JSON object (just like the `JSONEachRow` format does).

Reply:
if user produce `multiple JSON object` in one kafka message, and if users doesn't setup the JSONEachRow on our end, we will still treat them as one row instead of multiple since we can't decode it anyway.

yes, `one_message_per_row` is for sink and it is tricky but i was thinking when we query it back by using the same external stream, we still have the insights it is one message per row, although there may be others writing multiple rows in to the same kafka topic for which we can say it is users setup problem not us.

Reply:
Or we have a setting which allows user to use the optimization always. Like in kv stream, we have a `use_approximate_count`, when users query like `SELECT count() FROM table(kafka_topic) SETTINGS use_approximate_count=true`, we always fallback to this optimization which could be very helpful for users who really just like to know how many messages there in the topic for now.

Reply:
For this case, we can choose to use the same setting or choose another one like `count_kafka_message=true` which use this optimization always.

Reply:
I am thinking maybe we are overthinking for this issue.

There is already a setting called `optimize_trivial_count_query`. I suggest, let's just close that PR w/o merging it. And let the users set `optimize_trivial_count_query` to `false` if they find the optimisation does not work for them?

Reply:
After thinking about it a bit, I added `one_message_per_row` check and didn't add any new settings. I can't think of any other cases that people want to use message count as the count value for now. Let's keep it simple, and only add new settings when needed.